### PR TITLE
fix(ble): fix crash and reconnect bugs in iOS BLE stack

### PR DESCRIPTION
**Pull request by jdogg172**
## Summary

Fixes 5 crash-level bugs in the iOS BLE reconnection stack found during a full audit of the CoreBluetooth transport layer.

### BLETransport.swift
- **Force-unwrap crash**: Replaced `UUID(uuidString: device.identifier)!` with a `guard let` that throws `AccessoryError.connectionFailed` for malformed/empty identifiers
- **No reconnect on most errors**: `handlePeripheralDisconnectError` only retried on CBError codes 6 & 7; added retry for `.connectionFailed` (code 4) and all other codes
- **Duplicate `.deviceLost` events**: Removed double call to `connectionDidDisconnect` — `BLEConnection.disconnect()` already calls it internally
- **Background restore does not refresh config**: `handleWillRestoreState` `.connected` branch had `wantConfig/wantDatabase/versionCheck = false`; fixed to `true`

### BLEConnection.swift
- **Hanging async stream**: `getPacketStream()` overwrote `connectionStreamContinuation` without finishing the old one; added `finish()` before overwrite

## Tests
Added `MeshtasticTests/BLEReconnectTests.swift` covering UUID validation, RSSI boundary values, state restoration flag, and transport type invariants.

### CODE REVIEW

File: MeshtasticTests/BLEReconnectTests.swift [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+			let id = UUID()
+			let device = Device(
+				id: id,
+				name: "Radio",
+				transportType: .ble,
+				identifier: ""
+			)
+			// Device construction itself never crashes — the guard-let is inside BLETransport.connect(to:).
+			// This test validates the Device model accepts the value safely.
+			#expect(device.identifier == "")
```
[the remark this answers is not in the document]
Comment (jdogg172) [reply MeshtasticTests/BLEReconnectTests.swift]: Fixed: renamed to emptyIdentifierStoredSafelyInDeviceModel with an accurate comment explaining that the guard-let lives in BLETransport.connect(to:), not the Device model. The Foundation behaviour the fix relies on is now tested by the separate invalidUUIDStringsReturnNilFromFoundation parameterized test.

File: MeshtasticTests/BLEReconnectTests.swift [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+
+		/// Exact boundary -85: should be .weak (not .normal).
+		@Test func rssiAtWeakNormalBoundary() {
+			#expect(device(rssi: -85).getSignalStrength() == .weak)
+		}
+
+		/// One above the boundary (-84): should be .normal.
+		@Test func rssiJustAboveWeakNormalBoundary() {
+			#expect(device(rssi: -84).getSignalStrength() == .normal)
+		}
```
[the remark this answers is not in the document]
Comment (jdogg172) [reply MeshtasticTests/BLEReconnectTests.swift]: Fixed: removed the 4 boundary-value tests (-64, -65, -84, -85) that duplicated DeviceTests.signalStrength in ConnectViewTests.swift. Only the extremeRSSIValuesDoNotCrash test remains — it covers values (Int.min, 0) outside the range tested by the existing parameterized suite.

File: MeshtasticTests/BLEReconnectTests.swift [hunk 3]
```diff
@@ -0,0 +1,227 @@
+import Foundation
+import Testing
+
+@testable import Meshtastic
+
+// MARK: - BLE Reconnect Bug-Fix Tests
+//
+// These tests cover the fixes made in fix/ble-reconnect-crash:
```
[the remark this answers is not in the document]
Comment (jdogg172) [reply MeshtasticTests/BLEReconnectTests.swift]: Fixed: removed the branch name from the header. The comment now reads 'These tests cover the BLE reconnect crash fixes:' with no reference to the branch.

File: Meshtastic/Accessory/Transports/Bluetooth Low Energy/BLETransport.swift [hunk 4]
```diff
@@ -265,9 +268,14 @@ actor BLETransport: Transport {
...
+				Logger.transport.error("🛜 [BLETransport] Disconnected with CBError code: \(cbError.code.rawValue) - \(cbError.localizedDescription)")
+				shouldReconnect = true
 			default:
-				// Fallback for other CBError codes
+				// Fallback for other CBError codes — treat as transient to avoid permanent hangs
 				Logger.transport.error("🛜 [BLETransport] Disconnected with CBError code: \(cbError.code.rawValue) - \(cbError.localizedDescription)")
+				shouldReconnect = true
 			}
 		case let otherError:
 			Logger.transport.error("🛜 [BLETransport] Disconnected with non-CBError: \(otherError.localizedDescription)")
```
[the remark this answers is not in the document]
Comment (jdogg172) [reply Meshtastic/Accessory/Transports/Bluetooth Low Energy/BLETransport.swift:281]: Fixed: shouldReconnect = true now set in the non-CBError (otherError) branch as well, so all unrecognised CoreBluetooth errors trigger a reconnect attempt rather than permanently stopping reconnection.

File: Meshtastic/Accessory/Transports/Bluetooth Low Energy/BLETransport.swift [hunk 5]
```diff
@@ -277,12 +285,15 @@ actor BLETransport: Transport {
 			Logger.transport.debug("🛜 [BLETransport] Error while connecting. Resuming connection continuation with error.")
 			continuation.resume(throwing: error)
 			self.connectContinuation = nil
+			// Still notify the transport so state resets cleanly
+			connectionDidDisconnect(fromPeripheral: peripheral)
```
[the remark this answers is not in the document]
Comment (jdogg172) [reply Meshtastic/Accessory/Transports/Bluetooth Low Energy/BLETransport.swift]: Fixed: removed the connectionDidDisconnect(fromPeripheral:) call from the connectContinuation branch. connect(to:)'s own catch block already calls it after the continuation throws, so calling it here was the source of the duplicate .deviceLost events. Added a comment explaining the invariant.

**Comment by garthvh** [comment]
The tests seem to be mostly testing swift, all of this looks like AI generated code done in an hour, I am not seeing a real fixes here.  Have you tested this code on devices at all?

File: Meshtastic/Accessory/Transports/Bluetooth Low Energy/BLEConnection.swift [hunk 6]
```diff
@@ -179,6 +179,9 @@ actor BLEConnection: Connection {
 	
 	func getPacketStream() -> AsyncStream<ConnectionEvent> {
 		AsyncStream<ConnectionEvent> { continuation in
+			// Finish any previous stream so its consumer's `for await` loop terminates cleanly
+			// instead of hanging indefinitely on the abandoned continuation.
+			self.connectionStreamContinuation?.finish()
```
[the remark this answers is not in the document]
Comment (jake-b) [reply Meshtastic/Accessory/Transports/Bluetooth Low Energy/BLEConnection.swift:184]: The other connection types (TCP, Serial) have a termination handler for their implementations, so I agree one should be added to BLEConnection.

The proposed termination handler here is a little more than we have in the other Connection types.  This one checks for the `AsyncThrowingStream.Continuation.Termination` parameter, which can be `.finsihed(Failure?)` or `.cancelled`.   The other handlers don't make this distinction, and simply clean up when the continuation finishes or is cancelled.

The only place were we call .finish is in the disconnect() function itself, so its circular to call disconnect when the connection stream finishes.

Recommend adding the parameter and the `guard case .cancelled ` to the handlers in the other connection types as well.

File: Meshtastic/Accessory/Transports/Bluetooth Low Energy/BLETransport.swift [hunk 7]
```diff
@@ -413,7 +424,7 @@ actor BLETransport: Transport {
 				Logger.transport.error("🛜 [BLE] Peripheral Connection found and state is connected setting this connection as the activeConnection.")
 				let connectTask = Task { @MainActor in
 					// In this case we need a full reconnect, so do the wantConfig, wantDatabase, and versionCheck
-					try await AccessoryManager.shared.connect(to: device, withConnection: restoredConnection, wantConfig: false, wantDatabase: false, versionCheck: false)
+					try await AccessoryManager.shared.connect(to: device, withConnection: restoredConnection, wantConfig: true, wantDatabase: true, versionCheck: true)
```
Comment (jake-b) [inline Meshtastic/Accessory/Transports/Bluetooth Low Energy/BLETransport.swift:427]: The whole point of the handleWillRestoreState is that it des not do a full wantConfig, wantDatabase, and versionCheck.

This handler is called when the app is terminated in the background.  CoreBluetooth maintains the connection to the device, so the radio never knows the app was even killed.  If a new packet comes in, or the user launches the app before any new BLE traffic, the app restores itself.  No state has changed from when the app was terminated, so the app can resume with its current state.  The previous wantConfig and database are already in the database.

We should only need a full wantConfig/wantDatabase when the app is in a state where it has missed a packet (ie first connect, or an inability to restore and handle a packet, etc.

I think the comment might be wrong and that mislead the AI?

File: Meshtastic/Accessory/Transports/Bluetooth Low Energy/BLETransport.swift [hunk 8]
```diff
@@ -277,12 +285,15 @@ actor BLETransport: Transport {
...
 		} else if let activeConnection = self.activeConnection {
-			// Inform the active connection that there was an error and it should disconnect
-			Logger.transport.debug("🛜 [BLETransport] Error while connecting. Disconnecting the active connection.")
+			// Inform the active connection that there was an error and it should disconnect.
+			// BLEConnection.disconnect(withError:shouldReconnect:) internally calls connectionDidDisconnect,
+			// so we must NOT call it again here — doing so would emit duplicate .deviceLost events.
+			Logger.transport.debug("🛜 [BLETransport] Error while connected. Disconnecting the active connection.")
 			Task {
 				try? await activeConnection.disconnect(withError: error, shouldReconnect: shouldReconnect)
-				await self.connectionDidDisconnect(fromPeripheral: peripheral)
```
Comment (jake-b) [inline Meshtastic/Accessory/Transports/Bluetooth Low Energy/BLETransport.swift:285]: I think this one is okay, because yes, the call to .disconnect(withError:shouldReconnect:) does call connectionDidDisconnect already, so this is duplicative.

**Comment by garthvh** [comment]
Yeah I don't think I am going to merge any of these.

**Review by jake-b — COMMENTED** [review]
I made some comments!
